src/watch.js

```diff
@@ -238,32 +240,34 @@ class Watch {
     };
 
     /** Calls onError() and closes the stream. */
-    const sendError = function(errMessage) {
+    const sendError = function(err) {
       if (currentStream) {
         currentStream.unpipe(stream);
         currentStream.end();
         currentStream = null;
       }
       isActive = false;
       stream.end();
-      Firestore.log(
-        'Watch.onSnapshot',
-        'Invoking onError with: %s',
-        errMessage
-      );
-      onError(new Error(errMessage));
+      Firestore.log('Watch.onSnapshot', 'Invoking onError: ', err);
+      onError(err);
     };
 
     /**
-     * If the stream was previously healthy, we'll retry once and then give up.
+     * Re-opens the stream unless the specified error is considered permanent.
+     * Clears the change map.
      */
     const maybeReopenStream = function(err) {
-      if (isActive && isHealthy) {
+      if (isActive && (!err || !isPermanentError(err))) {
         Firestore.log('Watch.onSnapshot', 'Stream ended, re-opening');
         request.addTarget.resumeToken = resumeToken;
         changeMap.clear();
+
+        if (err && isResourceExhaustedError(err)) {
+          self._backoff.resetToMax();
+        }
+
         resetStream();
-      } else if (err) {
+      } else {
```

Comment:
why do we no longer check "if (err)" ?  Calling sendError() with null / undefined seems suspicious (it'll call onError() with no error, etc.)

Reply:
I fixed this by a) always calling this method with an error and b) only invoking the onError callback if we are still active.